src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,414 @@
+
+/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */
+
+/* fm-icon-container.h - the container widget for file manager icons
+
+   Copyright (C) 2024 Luke <lukefromdc@hushmail.com>
+
+   The Mate Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Library General Public License as
+   published by the Free Software Foundation; either version 2 of the
+   License, or (at your option) any later version.
+
+   The Mate Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with the Mate Library; see the file COPYING.LIB.  If not,
+   write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+
+   Author: <lukefromdc@hushmail.com>
+*/
+
+#include <config.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <X11/Xatom.h>
+#include <gtk/gtk.h>
+#include <gdk/gdkx.h>
+#include <glib/gi18n.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <eel/eel-glib-extensions.h>
+#include <eel/eel-gtk-extensions.h>
+#include <eel/eel-vfs-extensions.h>
+
+#include <libcaja-private/caja-directory-background.h>
+#include <libcaja-private/caja-global-preferences.h>
+
+#include "fm-desktop-icon-view.h"
+#include "fm-desktop-wayland-bg-dialog.h"
+
+
+#ifdef HAVE_WAYLAND
+
+GSettings   *background_settings;
+GtkWidget   *filebutton;
+GtkWidget   *preview;
+GtkWidget   *preview_image;
+const gchar *filename;
+GdkRGBA      color1, color2;
+const gchar  *primary_color_str,  *secondary_color_str, *shading_type;
+
+static void
+update_preview()
+{
+    static GtkCssProvider *provider;
+    gchar           *css;
+    GString         *string;
+
+    if (!provider)
+        provider =  gtk_css_provider_new ();
```

Comment:
```g_object_unref (provider);```

Reply:
This is now declared statically in update_preview() so it loads only once,  retains its value between preview updates, and isn't on the heap so it does not leak.Static variables are given space once when the program loads and deallocated when the program terminates according to what I was able to find online.